mllib/src/main/scala/org/apache/spark/ml/tuning/SlidingWindowCrossValidator.scala

```diff
@@ -0,0 +1,261 @@
+
+  /**
+   * Param for index column name.
+   * Default: "index"
+   * @group param
+   */
+  val indexCol: Param[String] = new Param[String](this, "indexCol", "index column name")
+
+  setDefault(indexCol, "index")
+
+  /** @group getParam */
+  def getIndexCol: String = $(indexCol)
+}
+
+private [tuning] object SlidingWindowCrossValidator {
+  def split(dataset: DataFrame, firstCutoffIndex: Long, windowSize: Long, numFolds: Int,
+    indexCol: String): Array[(DataFrame, DataFrame)] = {
+    (0 until numFolds).map { idx =>
+      val trainingUntil = firstCutoffIndex + idx * windowSize
+      val validationUntil = firstCutoffIndex + (idx + 1) * windowSize
+      val trainingSet = dataset.filter(dataset(indexCol) < trainingUntil)
+      val validationSet = dataset.filter(
+        dataset(indexCol) >= trainingUntil && dataset(indexCol) < validationUntil)
+      (trainingSet, validationSet)
+    }
+    .toArray
+  }
+}
+
+
+/**
+ * :: Experimental ::
+ * Sliding window cross validation.
+ *
+ * [[SlidingWindowCrossValidator]] provides a cross-validation method by splitting the data set by
+ * a cutoff index. The cutoff index is used to separate training and validation data set. Rows
+ * smaller than the cutoff index go to the training set and rows larger than it go to the validation
+ * set. This splitting method is useful for time-sensitive problem like fraud detection and stock
+ * price prediction.
+ *
+ * 3 parameters govern the splitting: numFolds is the number of splits, firstCutoffIndex denotes the
+ * the cutoff index of the first training data, and windowSize defines the index range of a
+ * validation set. In particular, training data is constructed with index range in [-inf, firstCutoffIndex
+ * + i * windowSize) and validation data is constructed with index range in [firstCutoffIndex +
+ * i * windowSize, firstCutoffIndex + (i + 1) * windowSize).
+ */
+@Experimental
+class SlidingWindowCrossValidator(override val uid: String)
+  extends Estimator[SlidingWindowCrossValidatorModel]
+  with SlidingWindowCrossValidatorParams with Logging {
+
+  def this() = this(Identifiable.randomUID("swcv"))
+
+  private val f2jBLAS = new F2jBLAS
+
+  /** @group setParam */
+  def setEstimator(value: Estimator[_]): this.type = set(estimator, value)
+
+  /** @group setParam */
+  def setEstimatorParamMaps(value: Array[ParamMap]): this.type = set(estimatorParamMaps, value)
+
+  /** @group setParam */
+  def setEvaluator(value: Evaluator): this.type = set(evaluator, value)
+
+  /** @group setParam */
+  def setNumFolds(value: Int): this.type = set(numFolds, value)
+
+  /** @group setParam */
+  def setFirstCutoffIndex(value: Long): this.type = set(firstCutoffIndex, value)
+
+  /** @group setParam */
+  def setWindowSize(value: Long): this.type = set(windowSize, value)
+
+  /** @group setParam */
+  def setIndexCol(value: String): this.type = set(indexCol, value)
+
+  override def fit(dataset: DataFrame): SlidingWindowCrossValidatorModel = {
```

Comment:
There is a lot of duplication between this and `CrossValidator`; perhaps we can think about subclassing here (i.e. defining a splitting method which defaults to `MLUtils.kFold` in `CrossValidator` and is overridden by `SlidingWindowCrossValidator.split` here)